A42-xds-ring-hash-lb-policy.md

```diff
@@ -0,0 +1,580 @@
+
+## Abstract
+
+We are adding support for hash-based load balancing configured via xDS.
+
+## Background
+
+There are two features that xDS supports that, in concert, allow
+hash-based load balancing: setting a hash policy in the route that
+determines the hash to be used for each request, and configuring a
+load balancing policy that uses that hash.
+
+xDS supports many different types of hash policies.  gRPC will support
+only a subset of these, but it will do so in a forward-compatible manner
+that will allow us to support additional hash policies in the future if
+needed.
+
+xDS supports two hash-based LB policies, `RING_HASH` and `MAGLEV`.  For
+now, gRPC will support only `RING_HASH`, although we could add support
+for `MAGLEV` in the future.
+
+### Related Proposals: 
+
+This proposal builds on earlier work described in the following gRFCs:
+* [gRFC A27: xDS-based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md)
+* [gRFC A28: xDS-based Traffic Splitting and Routing](https://github.com/grpc/proposal/blob/master/A28-xds-traffic-splitting-and-routing.md)
+* [gRFC A31: xDS-based Timeout and Config Selector](https://github.com/grpc/proposal/blob/master/A31-xds-timeout-support-and-config-selector.md)
+* [gRFC A37: xDS Aggregate and Logical DNS Clusters](https://github.com/grpc/proposal/blob/master/A37-xds-aggregate-and-logical-dns-clusters.md)
+
+## Proposal
+
+As mentioned above, gRPC will support two xDS features: setting a hash
+policy in the route, and configuring the `RING_HASH` LB policy.
+
+### Hash Policy
+
+This section explains how gRPC will support xDS-configured hash policies.
+
+#### xDS API Fields
+
+In xDS, the `RouteAction` message has a [`hash_policy`
+field](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/route/v3/route_components.proto#L1046)
+that specifies the policy used to compute the hash for each request that
+uses the route.  This is a `repeated` field, so it specifies a list of
+hash policies.  Each hash policy is evaluated individually, and the
+combined result is used to determine the request's hash.  The method of
+combination is deterministic, such that identical lists of hash policies
+will produce the same hash.  Since a hash policy examines specific parts of
+a request, it can fail to produce a hash (e.g., if the hashed header is not
+present).  If (and only if) all configured hash policies fail to generate a
+hash, a random hash will be used for the request, which (assuming a
+hash-based LB policy is used) will result in picking a random endpoint
+for the request.  If a hash policy has the `terminal` attribute set to true,
+and the policy does generate a result, then all subsequent hash policies
+are skipped.
+
+These semantics allow gRPC to support only a subset of xDS hash policy
+types in a forward-compatible way.  For any hash policy type that gRPC
```

Comment:
Somewhere we should probably note that this behavior differs from Envoy. https://github.com/envoyproxy/envoy/blob/c5833f2149cebec9fcdc8de0aec4bf425e0781d3/source/common/http/hash_policy.cc#L214-L215

Reply:
Done.